### ISSUE TITLE
Function sequence error (0) while inserting NULL value into NULLable VARCHAR/NVARCHAR(MAX) Column

### ISSUE THREAD

**Issue by gizmo93** [mkleehammer/pyodbc#450]
### Environment

To diagnose, we usually need to know the following, including version numbers.  On Windows, be
sure to specify 32-bit Python or 64-bit:

- Python: 3.5.4 64-bit
- pyodbc: 4.0.24
- OS: Windows 10
- DB: SQL Server 2017
- driver: ODBC Driver 17 for SQL Server

### Issue

Expected behavior: Insert NULL into nullable varchar/nvarchar(max) column with fast_executemany=True.
What happens: pyodbc.Error: Function sequence error (0) (SqlParamData)

Reproduction:

```
CREATE TABLE TestTable
(
	LongerContent NVARCHAR(MAX) NULL,
)
```

Works:

```
sql = """
INSERT INTO TestTable (LongerContent) VALUES (?)
"""

params = (
    ('ajhdiwopwqlwdlsdlpwdlowde',), (None,),
)


con = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=server;DATABASE=db;Trusted_Connection=yes;')
cursor = con.cursor()
cursor.fast_executemany=False
cursor.executemany(sql, params)
con.commit()
```

Results in Function sequence error (0):

```
import pyodbc

sql = """
INSERT INTO TestTable (LongerContent) VALUES (?)
"""

params = (
    ('ajhdiwopwqlwdlsdlpwdlowde',), (None,),
)


con = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=server;DATABASE=db;Trusted_Connection=yes;')
cursor = con.cursor()
cursor.fast_executemany=True
cursor.executemany(sql, params)
con.commit()
```

so it seems to be a problem with the fast_executemany option.

ODBC Trace:
[SQL.LOG](https://github.com/mkleehammer/pyodbc/files/2317693/SQL.LOG)

**Comment by v-chojas** [comment]
[The PR to fix this](https://github.com/mkleehammer/pyodbc/pull/425) was merged after .24 was released. Try the master branch.

**Comment by gizmo93** [comment]
Hey v-chojas,
thank you! Would it be possible to do a release including that bugfix?

**Comment by gordthompson** [comment]
@v-chojas  - Possibly related:

Using the current master branch and Python 3.6 the test `test_executemany_dae_0` in tests3/sqlservertests.py fails under Windows (msodbcsql17.dll 17.02.0000) but does not fail under Linux (libmsodbcsql-17.2.so.0.1 17.02.0000, unixODBC 2.3.7).

**Comment by tigerhawkvok** [comment]
@v-chojas Now giving a MemoryError instead!

**Comment by mkleehammer** [comment]
FYI - I'm still seeing the function sequence error on Windows.  However, a memory overwrite could certainly behave differently on different machines.

**Comment by gordthompson** [comment]
@v-chojas ( cc: @v-makouz , @mkleehammer  )

Further to [my comment above](https://github.com/mkleehammer/pyodbc/issues/450#issuecomment-417022316) and [your comment in issue 432](https://github.com/mkleehammer/pyodbc/issues/432#issuecomment-420781786)

When `test_executemany_dae_0` in "tests3/sqlservertests.py" succeeds on Linux we see `SQLExecute` returning with `SQL_NEED_DATA`, followed by `SQLParamData` returning with `SQL_NEED_DATA`, followed by `SQLPutData`

[Linux_ok.log](https://github.com/mkleehammer/pyodbc/files/2385922/Linux_ok.log), line 4120

```
[ODBC][1952][1537029166.431837][SQLExecute.c][187]
        Entry:
            Statement = 0x29aa270
[ODBC][1952][1537029166.432641][SQLExecute.c][357]
        Exit:[SQL_NEED_DATA]
[ODBC][1952][1537029166.432721][SQLParamData.c][166]
        Entry:
            Statement = 0x29aa270
            Value = 0x7ffef68d8fa0
[ODBC][1952][1537029166.432734][SQLParamData.c][362]
        Exit:[SQL_NEED_DATA]
            Value = 0x29ae680
[ODBC][1952][1537029166.432770][SQLPutData.c][148]
        Entry:
            Statement = 0x29aa270
            Data = 0x7f6b27b63690
            StrLen = 0
[ODBC][1952][1537029166.432784][SQLPutData.c][315]
        Exit:[SQL_SUCCESS]
```

However, when it fails on Windows we see `SQLExecute` returning with `SQL_NEED_DATA`, followed by `SQLParamData` returning with `SQL_NEED_DATA`, followed by another `SQLParamData` call ...

[Windows_fail.log](https://github.com/mkleehammer/pyodbc/files/2385924/Windows_fail.log), line 5572

```
sqlservertests  238c-5a8    ENTER SQLExecute 
        HSTMT               0x000000000008D740

sqlservertests  238c-5a8    EXIT  SQLExecute  with return code 99 (SQL_NEED_DATA)
        HSTMT               0x000000000008D740

sqlservertests  238c-5a8    ENTER SQLParamData 
        HSTMT               0x000000000008D740
        PTR *              0x00000000002ADDC8

sqlservertests  238c-5a8    EXIT  SQLParamData  with return code 99 (SQL_NEED_DATA)
        HSTMT               0x000000000008D740
        PTR *              0x00000000002ADDC8

sqlservertests  238c-5a8    ENTER SQLParamData 
        HSTMT               0x000000000008D740
        PTR *              0x00000000002ADDC8

sqlservertests  238c-5a8    EXIT  SQLParamData  with return code -1 (SQL_ERROR)
        HSTMT               0x000000000008D740
        PTR *              0x00000000002ADDC8

        DIAG [HY010] [Microsoft][ODBC Driver 17 for SQL Server]Function sequence error (0) 
```

... which is the "Function sequence error".

**Comment by v-chojas** [comment]
@gordthompson  Yes we are aware of the cause, and will be working on a solution to it.

**Comment by mkleehammer** [comment]
This seems pretty serious, so I'm going to wait for this before making the next build.

**Comment by v-makouz** [comment]
@gizmo93
I've made a pull request for a fix for this error. It seems that in python 3 the parameters are not bytes, but Unicode objects and to allow them to be used by SQLPutData I need to convert them into a bytes object.

Tested on Windows 10 and Ubuntu 17.04, I also ran sqlservertests.py, mysqltests.py and pgtests.py using Python 2 and 3 and didn't see any new errors.

https://github.com/mkleehammer/pyodbc/pull/467

@mkleehammer Let me know if you need me to make any further changes

**Comment by mkleehammer** [comment]
Merged in fix from v-makouz in 4.0.25
